### ISSUE TITLE
Consider global season variant

### ISSUE THREAD

**Issue by sebamarynissen** [memo33/sc4pac-tools#16]
I'm currently working on a mod that makes all trees made by Girafe semi-seasonal. With this I mean that depending what file you have in your plugins (z_fall, z_winter, z_summer, z_snow), the trees have your preferred appearance, regardless of whether they were actually planted as seasonal, evergreen or everwinter, inspired by the approach used in the [Central European Tree
 Controller](https://community.simtropolis.com/files/file/31419-central-european-tree-controller/).

I implemented the proof of concept today and hope to release it somewhere this week after some polishing. Obviously I also plan to write metadata for it so that it can be installed with sc4pac, which is especially useful because sc4pac makes it dead easy to switch between seasons by the use of variants.

While I was thinking about this, I was wondering whether it makes sense to introduce a global season preference. That would allow me to change both the CETC season and the season of my Girafe patch at the same time. I also plan on making a similar mod for other seasonal props and flora (for example cycledoggs's trees), so it would be nice to be able to switch seasons all at once instead of per variant.

I was thinking of the following possible values
 - `fall`, but could be called autumn as well
 - `winter`
 - `snow` This could default to the winter variant if a package does not define a specific snow variant
 - `spring` This could default to the summer variant if the package does not define it
 - `summer` This could be the global default if a package does not define fall or winter. Could also be named evergreen/static, or both could exist (where evergreen is global default then)
 - `seasonal` If this is not defined, then default to evergreen

Something to be aware of though is that a value of seasonal might actually be incompatible with any of the other values. For example, the CETC does allow you to switch between summer, winter and fall, but *not* between those and seasonal. That's actually already the case now with the CETC, but if this would be a global variant, it becomes something users need to be aware of.

I know that this is kind of a niche problem, but I think it is worth considering. What's your opinion, @memo33 ?

An example of what the metadata could look like for such a package:
```yaml
 - variant: { season: summer }
   assets:
     - assetId: girafe-patch
       include: [ "_SUMMER\.dat" ]
 - variant: { season: fall }
   assets:
     - assetId: girafe-patch
       include: [ "_FALL\.dat" ]
 - variant: { season: winter }
   assets:
     - assetId: girafe-patch
       include: [ "_WINTER\.dat" ]
 - variant: { season: snow }
   assets:
     - assetId: girafe-patch
       include: [ "_SNOW\.dat" ]
 - variant: { season: seasonal }
   assets: []
```

**Comment by memo33** [comment]
The main problems I see with this that
- switching between already installed variants is not allowed by some packages
- some packages support only a subset of the variants.

Such "incomplete" variants are not really intended currently. It leads to conflicts when installing a variant that is not supported by a package. There aren't any fallback mechanisms.

As switching variants would be done rather infrequently presumably, I'm not sure it's worth it to implement something special along those lines. Choosing variants case-by-case for each package is probably fine. If packages were _designed to work together_, they can use the same variant even, so you could re-use the variants of the CETC for example.

@Zasco might have some ideas on this topic.

**Comment by sebamarynissen** [comment]
Yeah, I understand it's kind of a niche thing. It's probably not worth the effort like you say. I really like the "static seasonal" strategy though of the CETC and I could imagine creators providing various versions of their lots for example. If there is a variant per lot, it can get quickly out of hand.

Still, it's unlikely that a creator would release 4 different versions of every lot they make: a seasonal, summer, fall and winter version. If they intend their lot to be used like that, they'll be better off creating it as seasonal and then refer to my mod that I'm about to create.

As I mentioned before, I plan on releasing a mod for Girafe's flora, but in the end the goal is to have it available for all seasonal flora. Would it be a good idea then to have a kind of "dummy" package that defines a variant, such as
```yaml
group: smf-16
name: flora
variants:
  - variant: { smf-16:flora:season: fall }
  - variant: { smf-16:flora:season: winter }
  - variant: { smf-16:flora:season: snow }
  - variant: { smf-16:flora:season: summer }
  - variant: { smf-16:flora:season: seasonal }
```
and then other packages I make can use those variants?

Alternatively I could also just put everything in one package. I'm doing the Girafe flora now, so I could release it once that is done, and then add other seasonal flora later on in the same package. That way a local variant is sufficient.

**Comment by Zasco** [comment]
> @Zasco might have some ideas on this topic.

This brings forth the why I'm working on _**SC4C**_, which you're not yet aware of _@sebamarynissen_. As you're contributing a lot to _**sc4pac**_, Ill add you to the related _Simtropolis_ PM so you can read what I already wrote about it.

**Comment by memo33** [comment]
> Would it be a good idea then to have a kind of "dummy" package that defines a variant, such as
> 
> ```yaml
> group: smf-16
> name: flora
> variants:
>   - variant: { smf-16:flora:season: fall }
>   - variant: { smf-16:flora:season: winter }
>   - variant: { smf-16:flora:season: snow }
>   - variant: { smf-16:flora:season: summer }
>   - variant: { smf-16:flora:season: seasonal }
> ```
> 
> and then other packages I make can use those variants?

Yes, I think. This sounds a bit like what was done with the [sfbt:essentials](https://memo33.github.io/sc4pac/channel/?pkg=sfbt:essentials) which define configurable tree families that are used by various SFBT plugins.

Whether it makes sense to put everything in one package, I don't know. It probably depends on how big it is and how it's used. I'm a bit skeptical about the amount of dependencies this would have. While technically sc4pac allows effortlessly installing large amounts of dependencies, I think that's still not desirable. Like, a package depending on all of Girafe's flora kind of defeats the point of splitting Girafe's flora into individual small packages in the first place. Maybe an approach similar to [nos17:essentials](https://memo33.github.io/sc4pac/channel/?pkg=nos17:essentials) could work:
> This package has no dependencies. However, if you are a creator using these prop families, note certain prop families *do* have additional cascading dependencies. Refer to the provided documentation for details.

**Comment by sebamarynissen** [comment]
I understand, but I'm using exemplar patching in my mod, so the only hard dependency will be your submenus dll. No need to have all of Girafe's flora, it will only patch what you have.

Anyway, I've decided to go with separate packages. One for Girafe's flora, and one for Cycledogg's flora - and perhaps more to come - that will all depend on a dummy package for the seasons. The reason is that this way, I can still write an override for Cycledogg's flora in the future (like how the Maxis HD tree override by T-Wrecks works). Users would then be able to choose: make Cycledoggs flora non-seasonal and choose which season, or override them with Girafe's flora, choosing for seasonal or a fixed season.

In any case it's a bit out of scope of sc4pac, but I'll definitely leverage it to make it easier to switch seasons. You can close this issue if you want, you'll see my PRs coming in when it's ready.

On a totally unrelated note: if you'd like to test the prototype that I made, feel free to ask!

**Comment by memo33** [comment]
Closing, as the main question is answered. As there's no fixed number of "season" variants every plugin can adhere to, this seems impractical on a global level.
